### DISCUSSION TITLE
Separate I/O schemas inconsistencies

### DISCUSSION THREAD

**Discussion by slafs** [fastapi/fastapi#10697] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from decimal import Decimal

import pytest
from fastapi import FastAPI, status
from pydantic import BaseModel, ConfigDict


def get_app(
    separate_input_output_schemas: bool,
    json_schema_serialization_defaults_required: bool,
):
    class Model(BaseModel):
        model_config = ConfigDict(
            json_schema_serialization_defaults_required=json_schema_serialization_defaults_required,
        )

        foo: int = 42
        bar: Decimal  # Decimal has different schemas for input (validation) and output (serialization)

    app = FastAPI(separate_input_output_schemas=separate_input_output_schemas)

    @app.post(
        "/",
        responses={
            status.HTTP_402_PAYMENT_REQUIRED: {
                "model": Model,
            },
        },
    )
    def root(model: Model) -> Model:
        return model

    return app


@pytest.mark.parametrize("json_schema_serialization_defaults_required", [True, False])
def test_suppress_separate_io_models(
    json_schema_serialization_defaults_required: bool,
):
    """
    Case #1 - this works fine, skip this function if you want.

    Setting separate_input_output_schemas to False
    suppresses generating separate schemas for input and output
    regardless of json_schema_serialization_defaults_required value.
    """
    app = get_app(
        separate_input_output_schemas=False,
        json_schema_serialization_defaults_required=json_schema_serialization_defaults_required,
    )
    schema = app.openapi()

    assert "Model" in schema["components"]["schemas"]
    assert "Model-Input" not in schema["components"]["schemas"]
    assert "Model-Output" not in schema["components"]["schemas"]


@pytest.mark.parametrize("json_schema_serialization_defaults_required", [True, False])
def test_separate_io_models(
    json_schema_serialization_defaults_required: bool,
):
    """
    Case #2

    Setting separate_input_output_schemas to True
    generates separate schemas for input and output
    regardless of json_schema_serialization_defaults_required value (because we have the Decimal field),
    but (bug?) generates the "Input" schema for the custom response.
    """

    app = get_app(
        separate_input_output_schemas=True,
        json_schema_serialization_defaults_required=json_schema_serialization_defaults_required,
    )

    schema = app.openapi()

    assert "Model" not in schema["components"]["schemas"]
    assert "Model-Input" in schema["components"]["schemas"]
    assert "Model-Output" in schema["components"]["schemas"]

    # Separate Input model for the request
    assert (
        "Model-Input"
        in schema["paths"]["/"]["post"]["requestBody"]["content"]["application/json"]["schema"]["$ref"]
    )

    # Separate Output model for the 200 response
    assert (
        "Model-Output"
        in schema["paths"]["/"]["post"]["responses"]["200"]["content"]["application/json"]["schema"]["$ref"]
    )

    # Separate Output model for the 402 response
    # This fails!
    assert (
        "Model-Output"
        in schema["paths"]["/"]["post"]["responses"]["402"]["content"]["application/json"]["schema"]["$ref"]
    )
```

### Description

I think I've spotted two issues regarding the separate Input and Output OpenAPI schemas  topic.

1. The `responses` -> _status_code_ -> `model` option for the given path operation generates an Input model instead of the Output one. See the `test_separate_io_models` test.
2. Suppressing separate models on FastAPI level works fine (see `test_suppress_separate_io_models`), but enabling it is more subtle than [the docs](https://fastapi.tiangolo.com/how-to/separate-openapi-schemas) are suggesting. 
   1. As of Pydantic v2.4.0 (https://github.com/pydantic/pydantic/pull/7275) fields with defaults are not required in the serialization schema by default. This means you have to explicitly opt-in to have such models generate different I/O schemas by setting `json_schema_serialization_defaults_required` config value to `True`. My example code doesn't actually show this, but you can observe it by commenting the `bar` field and noticing that only the `Model` schema does get generated when you set `json_schema_serialization_defaults_required` config to `False` and there is no `-Input` nor `-Output` models.
   2. Having a default value or not on a given field is not the only way to have different input and output schemas for the corresponding model. In Pydantic v2, one [can define](https://docs.pydantic.dev/latest/api/json_schema/#pydantic.json_schema.WithJsonSchema) a field that yields a different schema for validation (input) and serialization (output) without giving it a default value. One built-in type with such behaviour is `decimal.Decimal` class from the stdlib. However, [the same docs](https://fastapi.tiangolo.com/how-to/separate-openapi-schemas/#separate-openapi-schemas-for-input-and-output-or-not) seem to claim that the separation is only "depending on if they have default values".

### Operating System

macOS

### FastAPI Version

0.104.1

### Pydantic Version

2.5.1

### Python Version

Python 3.10.9

**Comment by slafs** [comment]
@Kludex sorry for pinging you directly. Is there any chance you could turn this into an issue?

**Comment by slafs** [comment]
Issue 1. is addressed in this PR https://github.com/tiangolo/fastapi/pull/11517

**Comment by slafs** [comment]
In case someone runs into this issue too:

> 1. The `responses` -> _status_code_ -> `model` option for the given path operation generates an Input model instead of the Output one. See the `test_separate_io_models` test.

If your model is being used only in the context of responses, you can force the mode (so the schema is being rendered properly) by adding a config option on your model, like this:
```py
     model_config = ConfigDict(
        json_schema_mode_override="serialization",
    )
```

**Comment by sjdemartini** [comment]
For what it's worth, it looks like Pydantic added an internal issue https://github.com/pydantic/pydantic/issues/8413 soon after https://github.com/pydantic/pydantic/pull/7275 was merged, which seems to be calling attention to your Issue (2) in your original post—fields with defaults confusingly end up as not required in the serialization schema.

**Comment by YuriiMotov** [comment, marked as answer]
This has been fixed 🎉
Works fine in current versions (FastAPI 0.115.12, Pydantic 2.11.5)
